### ISSUE TITLE
[Question] how rotation works 

### ISSUE THREAD

**Issue by studentdeng** [Flipboard/FLEX#62]
relate to #59 .

Could you explain why we should care about `status bar` and 

> our window shouldn't be the key window when this view controller is asked about the status bar ?

I think FLEX Windows is just a Floating Window.

**Comment by ryanolsonk** [comment]
Sure, I can take a stab at explaining this.

Status bar and rotation management somewhat are related because they are both view controller based. Previously, FLEX did an OK job of finding an appropriate view controller from the main app to query for status bar and rotation behavior. However, there were constantly bugs from this and I knew there had to be a better way. Prompted by #59, I investigated deeper into how the system handles these to see if there was any private API that could help us do better. 

It's difficult to add a window to the hierarchy and not affect the status bar appearance or rotation behavior. Let's start by walking through what happens when the system wants to update the status bar. I'm using [Hopper](http://www.hopperapp.com/) to disassemble UIKit (iOS 8.3). See http://www.objc.io/issue-19/debugging-case-study.html for a nice guide on how to set that up.

The first thing the system does is walk through the applications windows. It starts by calling `-[UIApplication _statusBarControllingWindow]` which passes through to `+[UIWindow _statusBarControllingWindow]`. The implementation of this method looks roughly like this:

```objc
+ (UIWindow *)_statusBarControllingWindow
{
    return [self _topVisibleWindowPassingTest:^BOOL(UIWindow *window) {
        return [window _canAffectStatusBarAppearance] && [window.screen _isMainScreen];
    }];
}
```

The method names here are descriptive enough to get an idea of what's going on. The system iterates through the windows from top to bottom until it finds one that is visible and says it can affect the status bar. The default implementation of `-[UIWindow _canAffectStatusBarAppearance]` looks like this:

```objc
- (BOOL)_canAffectStatusBarAppearance
{
    return ![self isInternalWindow];
}
```

So, if the FLEX window is visible and above the application's main window(s), it will be the one returned as the `_statusBarControllingWindow`. `UIViewController` has a class method for determining which view controller should ultimately be consulted for the status bar style. The method makes use of what we looked at above, and looks something like this:

```objc
+ (UIViewController *)_currentStatusBarStyleViewController
{
    UIWindow *window = [UIWindow _statusBarControllingWindow];
    return [[window rootViewController] _effectiveStatusBarStyleViewController];
}
```

When FLEX is only showing the toolbar, we want the underlying app to determine the status bar appearance. However, when FLEX presents a full screen modal view controller, we do want to take control of the status bar. To achieve this, `FLEXWindow` now overrides the private method `-_canAffectStatusBarAppearance` so it can choose whether it decides the status bar appearance or not.

Rotation is a little bit different because the different windows can rotate independently. When the device rotates, the root view controller of each window is asked for a `_viewControllerForSupportedInterfaceOrientations` and a `_viewControllerForRotation`. With FLEX, we intend to match the underlying app's rotation/orientation behavior. Therefore, when responding to the `shouldAutorotate` and `supportedInterfaceOrientations` methods in `FLEXExplorerViewController`, we look for a window from the main app and query it for the appropriate behavior.

That may have been a bit of a roundabout response, but hopefully it helps. Let me know if you have any further questions.
